logging/cef_formatter.go

```diff
@@ -0,0 +1,212 @@
+		timestampFormat = defaultTimestampFormat
+	}
+
+	hostname := f.HostName
+	if hostname == "" {
+		hostname = defaultHostName
+	}
+
+	logPrefix := f.CEFPrefixString
+	if logPrefix == "" {
+		logPrefix = defaultCEFLogStart
+	}
+
+	// timestamp host CEF:0
+	if f.ShouldAddSyslogPrefix {
+		b.WriteString(entry.Time.Format(timestampFormat))
+		b.WriteByte(' ')
+		b.WriteString(hostname)
+		b.WriteByte(' ')
+	}
+	b.WriteString(defaultCEFLogStart)
+
+	// |Device Vendor|Device Product|Device Version|Signature ID|Name|Severity|
+	f.appendCEFLogPiece(b, entry.Data[FieldKeyVendor])
+	f.appendCEFLogPiece(b, entry.Data[FieldKeyProduct])
+	f.appendCEFLogPiece(b, entry.Data[FieldKeyVersion])
+	f.appendCEFLogPiece(b, entry.Data[FieldKeyEventCode])
+
+	f.appendCEFLogPiece(b, entry.Message)
+	f.appendCEFLogPiece(b, severityByLevel(entry.Level))
+
+	b.WriteString(defaultMessageDivider)
+
+	// Extension
+
+	// actually, these fields should have only designated names according to the Extension Dictionary of CEF
+	extensionKeys := otherExtensionKeys(entry.Data)
+	for _, key := range extensionKeys {
+		f.appendKeyValue(b, key, entry.Data[key])
+	}
+
+	b.WriteByte('\n')
+	return b.Bytes(), nil
+}
+
+func otherExtensionKeys(data logrus.Fields) []string {
+	extensionKeys := make([]string, 0, len(data))
+	for k := range data {
+
+		if k != FieldKeyVendor && k != FieldKeyProduct && k != FieldKeyVersion &&
+			k != FieldKeyEventCode && k != FieldKeySeverity {
+
+			extensionKeys = append(extensionKeys, k)
+		}
+	}
+	return extensionKeys
+}
+
+func (f *CEFTextFormatter) appendCEFLogPiece(b *bytes.Buffer, value interface{}) {
+	b.WriteString(defaultMessageDivider)
+	f.appendValue(b, value)
+}
+
+func (f *CEFTextFormatter) appendKeyValue(b *bytes.Buffer, key string, value interface{}) {
+	preparedKey := prepareString(key)
+
+	if f.needsQuoting(preparedKey) {
+		preparedKey = fmt.Sprintf("%q", preparedKey)
+	}
+
+	b.WriteString(preparedKey)
+	b.WriteByte('=')
+	f.appendValue(b, value)
+	b.WriteByte(' ')
+}
+
+func (f *CEFTextFormatter) appendValue(b *bytes.Buffer, value interface{}) {
+	stringVal, ok := value.(string)
+	if !ok {
+		stringVal = fmt.Sprint(value)
+	}
+
+	stringVal = prepareString(stringVal)
+
+	// CEF doesn't define using quotes
+	if len(stringVal) == 0 {
+		b.WriteString(" ")
+	} else {
+		b.WriteString(stringVal)
+	}
+}
+
+func prepareString(value string) string {
+	stringVal := fmt.Sprint(value)
+
+	// is it a valid way to remove any \t\n even inside line?
+	stringVal = strings.TrimSpace(stringVal)
+	stringVal = strings.Replace(stringVal, "\n", " ", -1)
+	stringVal = strings.Replace(stringVal, "\t", " ", -1)
+	stringVal = strings.Replace(stringVal, `\`, `\\`, -1)
+	stringVal = strings.Replace(stringVal, "|", `\|`, -1)
+	stringVal = strings.Replace(stringVal, `=`, `\=`, -1)
+	return stringVal
+}
+
+func severityByLevel(level logrus.Level) int {
+	switch level {
+	case logrus.DebugLevel:
+		return 0
+	case logrus.InfoLevel:
+		return 0
```

Comment:
Did you mean `return 1`? Or, possibly, there is some reason to set the same value as for `DebugLevel`?